packages/lookit-initjspsych/src/utils.ts

```diff
@@ -1,132 +1,62 @@
+ * @returns On data update function.
+ */
+export const on_data_update = (responseUuid: string, userFunc?: UserFunc) => {
   return async function (data: DataCollection) {
-    const {
-      data: {
-        attributes: { exp_data },
-      },
-    } = await get(responseApiUrl);
+    const { attributes } = await Api.retrieveResponse(responseUuid);
+    const exp_data = attributes.exp_data ? attributes.exp_data : [];
 
-    await patch(responseApiUrl, true, {
-      id,
-      type: "responses",
-      attributes: {
-        exp_data: [...exp_data, data],
-      },
+    await Api.updateResponse(responseUuid, {
+      exp_data: [...exp_data, data],
     });
 
     // Don't call the function if not defined by user.
     if (typeof userFunc === "function") {
       userFunc(data);
     }
   };
-}
-
-export function on_finish(
-  responseApiUrl: string,
-  id: string,
-  exitUrl: string,
-  userFunc?: UserFunc,
-) {
-  /**
-   * Function that returns a function to be used in place of jsPsych's option
-   * "on_finish".  "userFunc" should be the user's implementation of
-   * "on_finish".  Since this is point where the experiment has ended, the
-   * function will set "completed" to true and overwrites all experiment data
-   * with the full set of collected data.  Once the user function has been
-   * ran, this will redirect to the study's exit url.
-   */
+};
+
+/**
+ * Function that returns a function to be used in place of jsPsych's option
+ * "on_finish". "userFunc" should be the user's implementation of "on_finish".
+ * Since this is point where the experiment has ended, the function will set
+ * "completed" to true and overwrites all experiment data with the full set of
+ * collected data. Once the user function has been ran, this will redirect to
+ * the study's exit url.
+ *
+ * @param responseUuid - Response UUID.
+ * @param userFunc - "on finish" function provided by the researcher.
+ * @returns On finish function.
+ */
+export const on_finish = (responseUuid: string, userFunc?: UserFunc) => {
   return async function (data: DataCollection) {
-    /**
-     * The on_data_update and on_finish functions aren't called as async
-     * functions.  This means that each function isn't completed before the
-     * next is ran. To handle this, we're going to abort the patch function
-     * in on_data_update.  This will cause a reliable error,
-     */
-    controller.abort("Writing final response data.");
+    const { exit_url } = window.chs.study.attributes;
 
-    await patch(responseApiUrl, false, {
-      id,
-      type: "responses",
-      attributes: {
-        exp_data: data.values(),
-        completed: true,
-      },
+    await Api.finish();
+    await Api.updateResponse(responseUuid, {
+      exp_data: data.values(),
+      completed: true,
     });
+    await Api.finish();
 
     // Don't call the function if not defined by user.
     if (typeof userFunc === "function") {
       userFunc(data);
     }
```

Comment:
Question: it looks like we're updating the response data and then calling the user function (if there is one). Would it be possible to switch the order of these two things, so that the user function is called before the final data update? The reason is that a common researcher use case for the `on_finish` function is to modify the trial/experiment data after the trial/experiment has ended.

Reply:
Yes.  I'll make that change in this PR.